### ISSUE TITLE
Latent security issues

### ISSUE THREAD

**Issue by aurelianito** [lief-project/LIEF#165]
We am trying to secure LIEF against malformed PE files. I was basing upon the code at `VectorStream::read`, that I understand that was renamed to `VectorStream::read_at` and made protected.

But now the code was changed to not throw a `LIEF::read_out_of_bound`, instead returning `null` (current code below). 

```
const void* VectorStream::read_at(uint64_t offset, uint64_t size) const {

  if (offset > this->size() or (offset + size) > this->size()) {
    size_t out_size = (offset + size) - this->size();
    LOG(ERROR) << "Can't read "
               << std::dec << size << " bytes at "
               << std::hex << std::showbase << offset
               << " (" << std::hex << (out_size) << " bytes out of bound)";
    //throw LIEF::read_out_of_bound(offset, size);
    return nullptr;
  }
  return this->binary_.data() + offset;
}
```

This is a nightmare, security wise, leaving lots of attack surface. If just a single check for `null` is forgotten then we can be owned. 

Can we go back to the old behavior where it throws an exception if we attempt to read out of bounds? If not possible we will be forced to use an old version, as it is impossible for us to check that every usage is doing the proper check.

BTW, we also found some other security issues in `PE\utils.cpp` (methods `is_pe` and `get_type`) and we had a patch based upon the original behavior for the read method.

**Comment by aurelianito** [comment]
This is one example of code that does not check for null 

```template<class T>
const T& BinaryStream::peek(void) const {
  const void* raw = this->read_at(this->pos(), sizeof(T));
  return *reinterpret_cast<const T*>(raw);
}```

BinaryStream.hpp lines 117-121.

**Comment by romainthomas** [comment]
Yes and actually ``read_array`` and ``peek_array`` will also throw an exception. See: https://github.com/lief-project/LIEF/commit/1f32b253295535de77f81997ac99c49fb8015c83

**Comment by aurelianito** [comment]
I have opened #167 with a bug fix for the possible security issues we have noticed in PE/utils.cpp

**Comment by romainthomas** [comment]
The PR has been merged.

**Comment by romainthomas** [comment]
If you are ok I'll close the issue
